vowpalwabbit/automl.cc

```diff
@@ -0,0 +1,917 @@
+      std::set<namespace_index> other_namespaces;
+      bin_text_read_write_fixed(model_file, (char*)&ns, sizeof(ns), read, msg, text);
+      bin_text_read_write_fixed(
+          model_file, reinterpret_cast<char*>(&other_ns_size), sizeof(other_ns_size), read, msg, text);
+      for (size_t j = 0; j < other_ns_size; ++j)
+      {
+        namespace_index other_ns;
+        bin_text_read_write_fixed(model_file, (char*)&other_ns, sizeof(other_ns), read, msg, text);
+        other_namespaces.insert(other_ns);
+      }
+      exclusions[ns] = other_namespaces;
+    }
+  }
+  else
+  {
+    exclusion_size = exclusions.size();
+    msg << "exclusion_size " << exclusion_size << "\n";
+    bin_text_read_write_fixed(
+        model_file, reinterpret_cast<char*>(&exclusion_size), sizeof(exclusion_size), read, msg, text);
+    for (const auto& ns_pair : exclusions)
+    {
+      namespace_index ns = ns_pair.first;
+      size_t other_ns_size = ns_pair.second.size();
+      bin_text_read_write_fixed(model_file, (char*)&ns, sizeof(ns), read, msg, text);
+      bin_text_read_write_fixed(
+          model_file, reinterpret_cast<char*>(&other_ns_size), sizeof(other_ns_size), read, msg, text);
+      for (const auto& other_ns : ns_pair.second)
+      { bin_text_read_write_fixed(model_file, (char*)&other_ns, sizeof(other_ns), read, msg, text); }
+    }
+  }
+
+  if (!read) msg << "_aml_lease " << lease << "\n";
+  bin_text_read_write_fixed(model_file, reinterpret_cast<char*>(&lease), sizeof(lease), read, msg, text);
+
+  if (!read) msg << "_aml_ips " << ips << "\n";
+  bin_text_read_write_fixed(model_file, reinterpret_cast<char*>(&ips), sizeof(ips), read, msg, text);
+
+  if (!read) msg << "_aml_lower_bound " << lower_bound << "\n";
+  bin_text_read_write_fixed(model_file, reinterpret_cast<char*>(&lower_bound), sizeof(lower_bound), read, msg, text);
+
+  if (!read) msg << "_aml_state " << state << "\n";
+  bin_text_read_write_fixed(model_file, reinterpret_cast<char*>(&state), sizeof(state), read, msg, text);
+}
+
+// config_manager is a state machine (config_manager_state) 'time' moves forward after a call into one_step()
+// this can also be interpreted as a pre-learn() hook since it gets called by a learn() right before calling
+// into its own base_learner.learn(). see learn_automl(...)
+interaction_config_manager::interaction_config_manager(size_t global_lease, size_t max_live_configs, uint64_t seed,
+    size_t priority_challengers,
+    float (*calc_priority)(const exclusion_config&, const std::map<namespace_index, size_t>&))
+    : global_lease(global_lease)
+    , max_live_configs(max_live_configs)
+    , seed(seed)
+    , priority_challengers(priority_challengers)
+    , calc_priority(calc_priority)
+{
+  random_state.set_random_state(seed);
+  exclusion_config conf(global_lease);
+  conf.state = Live;
+  configs[0] = conf;
+  scored_config sc;
+  scores.push_back(sc);
+}
+
+void interaction_config_manager::one_step(const multi_ex& ec)
```

Comment:
looks like `config_manager` is super generic and so it can be open to many different automl implementations, no?

Reply:
@rajan-chari I moved one_step() to the automl struct and put schedule(), config_oracle() etc in the config_manager base class. There are a few ways we could go about swapping out these functions. Since their inputs/outputs may vary based on config_manager type, algorithm used, etc... Im not certain it makes sense to swap in new function pointers (like what is done with calc_priority_empty/calc_priority_least_exclusion). Instead we could keep these functions generic (void with no inputs), and their behavior could change within a given implementation (with some if/else), based on some command-line argument.